code/modules/power/stirling.dm

```diff
@@ -0,0 +1,222 @@
+	construct_state = /decl/machine_construction/default/panel_closed
+	uncreated_component_parts = null
+	stat_immune = NOSCREEN | NOINPUT | NOPOWER // Mechanical machine and display.
+	connect_types = CONNECT_TYPE_REGULAR|CONNECT_TYPE_FUEL
+
+	var/obj/item/tank/stirling/inserted_cylinder
+
+	var/cycle_frequency = MAX_FREQUENCY/2
+	var/active = FALSE
+
+	var/max_power = 100000
+	var/genlev = 0
+	var/last_genlev
+
+	var/last_gen = 0
+	var/skipped_cycle = FALSE
+
+	var/sound_id
+	var/datum/sound_token/sound_token
+
+/obj/machinery/atmospherics/binary/stirling/Destroy()
+	QDEL_NULL(inserted_cylinder)
+	QDEL_NULL(sound_token)
+	. = ..()
+
+/obj/machinery/atmospherics/binary/stirling/Process()
+	..()
+
+	// Some temperature equilibrium between the gas lines.
+	var/air1_eq = air1.get_thermal_energy_change(air2.temperature)
+	var/air2_eq = air2.get_thermal_energy_change(air1.temperature)
+
+	var/heat_transfer = clamp(HEAT_TRANSFER*(air1.temperature - air2.temperature), min(air1_eq, air2_eq), max(air1_eq, air2_eq))
+	if(heat_transfer)
+		air1.add_thermal_energy(-heat_transfer)
+		air2.add_thermal_energy(heat_transfer)
+
+	if(!istype(inserted_cylinder))
+		return
+	if(!active)
+		return
+
+	var/datum/gas_mixture/working_volume = inserted_cylinder.air_contents
+
+	if((stat & BROKEN) || !air1.total_moles || !air2.total_moles || !working_volume?.total_moles)
+		stop_engine()
+		return
+	
+	// A rough approximation of a Stirling cycle with perfect regeneration e.g. Carnot efficiency.
+	var/working_heatcap = working_volume.heat_capacity()
+	var/line1_heatcap = air1.heat_capacity()
+	var/line2_heatcap = air2.heat_capacity()
+
+	// Bring the internal tank in thermal equilibrium with the hottest line. The temperature/pressure
+	// is kept at the maximum constantly.
+	var/equil_transfer
+
+	if(air1.temperature >= air2.temperature)
+		equil_transfer = (working_heatcap*line1_heatcap)/(working_heatcap + line1_heatcap)*(air1.temperature - working_volume.temperature)
+		air1.add_thermal_energy(-equil_transfer)
+	else
+		equil_transfer = (working_heatcap*line2_heatcap)/(working_heatcap + line2_heatcap)*(air2.temperature - working_volume.temperature)
+		air2.add_thermal_energy(-equil_transfer)
+
+	working_volume.add_thermal_energy(equil_transfer)
+
+	// The cycle requires a minimum temperature to overcome friction.
+	if(abs(air1.temperature - air2.temperature) < MIN_DELTA_T)
+		if(skipped_cycle) // Allow some leeway since networks can take some time to update.
+			stop_engine()
+		skipped_cycle = TRUE
+		return
+	skipped_cycle = FALSE
+
+	// Positive/negative Work from volume expansion/compression. Maximum volume is 1.5 tank volume.	The volume of the tank is not actually adjusted.
+	var/work_coefficient = working_volume.get_total_moles()*R_IDEAL_GAS_EQUATION*log(1.5)
+	var/work_done = work_coefficient*abs(air1.temperature - air2.temperature)*cycle_frequency
+
+	var/reversed = air1.temperature < air2.temperature ? 1 : -1
```

Comment:
This `reversed` var naming intuitively makes me think, that it will store boolean type, but it assigned with 1 or -1 which is confusing a bit.

Lower in the code, there is condition, that will be always true, because, as I can see, there is no other changes to this var. 1 and -1 both return TRUE in boolean equivalent.
```
if(reversed)
   air1_dq += 0.75*(max_power - power_generated)
else
   air2_dq += 0.75*(max_power - power_generated)
```
Also, what do you think about moving 1 and -1 to local #defines? I suppose, they must mean something

Reply:
Good catch. At some point ``reversed`` was a boolean, which is why this check was there, and I probably forgot to change it. I renamed the variable to ``heat_dir`` and swapped the sign so it makes more sense. ``heat_dir = 1`` corresponds to heat transfer from ``air_1`` to ``air_2``, and vice versa.